[[ Bug 18376 ]] Fix slow performance of trueWord chunk

runrevmark opened the pull request:
This patch corrects MCStringGetCodepointAtIndex which previously had assumed the input index was in codepoints, not in codeunits.

As an addendum, all uses of MCStringGetCodepointAtIndex have been vetted, and those which were fetching values at indicies > 0 now correctly process surrogates (i.e. skip the right number of codeunits).

runrevmark committed:
[[ Foundation ]] Fix incorrect mapping of indices in MCStringGetCodepointAtIndex

This patch removes the mapping of indices from MCStringGetCodepointAtIndex.
The mapping was unnecessary as the function is meant to treat the incoming
index as a code unit index.

runrevmark committed:
[[ Bug ]] Ensure the type command works correctly with surrogates

This patch fixes the 'type' command to correctly handle the case of
surrogate pairs.

runrevmark committed:
[[ Bug ]] Correct use of codepoints in TextFindNextParagraph

This patch corrects the use of codepoints in MCParagraph::TextFindNextParagraph
to ensure it skips the trailing surrogate correctly.

runrevmark committed:
[[ Bug ]] Eliminate unnecessary use of codepoints in resolve prop

This patch changes the use of GetCodepointAtIndex to GetCharAtIndex
in MCProperty::resolveprop since it is only looking for ASCII chars.

runrevmark committed:
[[ Bug 18376 ]] Fix use of GetCodepointAtIndex in MCLocaleWordBreakIteratorAdvance

This patch fixes the use of MCStringGetCodepointAtIndex in the trueWord
iterator - ensuring it skips the trailing surrogate correctly (if present).

runrevmark committed:
[[ Foundation ]] Remove unused MCChunkIterate and friends

This patch removes MCChunkIterate (which used MCStringGetCodepointAtIndex) and other
related functions which are no longer used anywhere.

runrevmark committed:
[[ Bug 18376 ]] Add release note

livecodeali commented:
@livecode-vulcan review ok b3d3a39

peter-b commented on engine/src/paragraf.cpp:
```diff
@@ -278,10 +278,19 @@ bool MCParagraph::TextIsPunctuation(codepoint_t p_codepoint)
+		
 		p_after++;
+		if (t_char > UNICHAR_MAX)
+			p_after++;
```
For consistency with the changes you've made to `exec-interface.cpp`, what about:

``` C++
p_after += t_char <= UNICHAR_MAX ? 1 : 2;
```

peter-b commented on libfoundation/src/foundation-locale.cpp:
```diff
@@ -1143,10 +1143,15 @@ bool MCLocaleWordBreakIteratorAdvance(MCStringRef self, MCBreakIteratorRef p_ite
+			
+			t_left_break++;
+			if (t_cp > UNICHAR_MAX)
+				t_left_break++;
```
Hmm... actually, is this a sufficiently common pattern that there ought to be an inline helper function?

``` C++
t_left_break += MCUnicodeCodepointNumChars(t_cp);
```

or something similar?

livecode-vulcan commented:
💙 review by @livecodeali ok b3d3a3989cdc39bcec8f7afd85ec39a9a5e563ee

peter-b commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -1581,31 +1581,31 @@ MC_DLLEXPORT_DEF
         self = self -> string;
     
+    /* Allow trailing null character */
+    MCAssert(p_index <= MCStringGetLength(self));
```
Just to be clear: it's okay for `MCStringGetCodepointAtIndex()` to retrieve the trailing nul as a codepoint? Why isn't it an error to try to read from beyond `MCString`'s explicitly-provided length?

runrevmark replied:
Because GetCharAtIndex and GetNativeCharAtIndex do - why they do is another question entirely.

peter-b reviewed (changes requested):
@runrevmark I think that adding an `inline` helper function would help avoid so many copies of the

``` C++
t_index++;
if (t_char > UNICHAR_MAX)
    t_index++;
```

snippet, and would be obvious from inspection.

runrevmark committed:
[[ Foundation ]] Add API to compute length of codepoint in codeunits

This patch adds MCUnicodeCodepointGetCodeunitLength() which returns
either 1 or 2, depending on whether the codepoint needs to be
represented as a surrogate pair.

Additionally, various places which advance by a codepoint have been
updated to use the new API.

livecodeali commented:
Alternatively, there could be a variant of `GetCodepointAtIndex` that also returned the increment as an out parameter.

peter-b reviewed (approved):
Looks good to me!

peter-b commented:
@livecode-vulcan review ok 696d275

livecode-vulcan commented:
💙 review by @peter-b ok 696d2759b00b02651e6f3bd28cd28d92283e0e3a

runrevmark commented:
Wait I think I missed one...

peter-b reviewed (approved):
@livecode-vulcan review ok fcfc251

peter-b commented:
Once more for @livecode-vulcan:
@livecode-vulcan review ok fcfc251

livecode-vulcan commented:
💙 review by @peter-b ok fcfc2519294d4cf0ed0b5ec9a89554b5c9776f10

livecode-vulcan commented:
😎 test success fcfc2519294d4cf0ed0b5ec9a89554b5c9776f10
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/212)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/216)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/212)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/212)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/216)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/212)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/212)
